### PR TITLE
Npc traffic spawning feature

### PR DESCRIPTION
**Pull request by johnMinelli**
The scenarios can be customized adding `vehicles` and `pedestrians` keys in order to spawn n NPC in the map
Additionally the environment now present a public variable `exclude_hard_vehicles` which allow to exclude the big vehicles (note also that the fixed camera position clip in some of those vehicles)

### CODE REVIEW

**Comment by praveen-palanisamy** [comment]
@johnMinelli ,
Could you please respond to and/or resolve the review comments so that we can make progress towards merging this PR?
The review comments are listed on this page above. 
Based on your [comment here](https://github.com/praveen-palanisamy/macad-gym/pull/68#issuecomment-1298683264), it may be the case that the comments are not showing up on your end. If you are unable to see them on this page too for some reason, please share a screenshot of this page including this comment in order for me to understand your view/interface and provide you relevant information in an alternative way.

**Comment by johnMinelli** [comment]
Hi, I would be happy to do it but I don't see any comment 🙂
image

**Review by praveen-palanisamy — CHANGES_REQUESTED** [review]
Thank you for letting me know. 
I am (re-)submitting partial review comments. Please address them to take this forward. 
Please let me know if this feedback is also not visible to you.

File: src/macad_gym/carla/config.json
Comment (johnMinelli) [reply src/macad_gym/carla/config.json:28]: yeah, this is more of a proposal since i wanted to make also the position customizable: the number (if present) allow to select which camera position between the default existing, to use. Do you think that the compliance is more important respect the feature or it is not necessary?

File: src/macad_gym/carla/config.json
Comment (johnMinelli) [reply src/macad_gym/carla/config.json:48]: i added the ability to use many different camera types

File: src/macad_gym/carla/config.json
Comment (praveen-palanisamy) [reply src/macad_gym/carla/config.json:28]: I see.  For identifying and specifying the camera or it's position, a camera ID (or position ID) can be used. 
The `camera-type` should be consistent with CARLA's camera type.

Comment (johnMinelli) [reply src/macad_gym/carla/config.json:28]: do you suggest to add a new parameter e.g. `camera-position` ?

Comment (praveen-palanisamy) [reply src/macad_gym/carla/config.json:28]: Yes, to add support for multiple cameras mounted on the car with configurable positions, adding `camera_position` sounds better than changing the `camera_type`.

Comment (johnMinelli) [reply src/macad_gym/carla/config.json:28]: Ok, I reviewed also the other comments but I will close them as soon as I will be able to fix them in the code

**Review by johnMinelli — COMMENTED** [review]
Sorry for the delay. I made the changes as you suggested, let me know if it is ok.

File: src/macad_gym/carla/config.json
Comment (praveen-palanisamy) [reply src/macad_gym/carla/config.json:49]: Okay but, what does the `ray` camera type refer to? 
CARLA doesn't have a `ray` camera.

File: src/macad_gym/carla/multi_env.py
Comment (praveen-palanisamy) [inline src/macad_gym/carla/multi_env.py:394]: This method is used to get a resource collision-free TCP socket that is guaranteed to be free/available. This is not meant to be a method to return a socket object given a particular port number.
Is there a particular reason you wanted to make this change to make the `port` number an argument? If not, please revert this change.

File: src/macad_gym/carla/multi_env.py
Comment (praveen-palanisamy) [inline src/macad_gym/carla/multi_env.py:409]: Please see comment above on the method definition and update this accordingly

File: src/macad_gym/carla/multi_env.py
Comment (praveen-palanisamy) [inline src/macad_gym/carla/multi_env.py:420]: (optional) Suggestion to use `logger.info(...)` instead of `print(...)`.
Also, since the python version in the spec is >3.6, consider using [f-strings](https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python) to make this more compact.

Example: `logger.info(f" 1. Port: {self._server_port} \n 2. Map: {self._server_map} \n 3. Binary: {SERVER_BINARY}")`.
(I used `\n` for the new line character instead of a pre-formatted multi-line string)

**Review by praveen-palanisamy — CHANGES_REQUESTED** [review]
@johnMinelli , Thank you for the updates! I have left a few minor comments to be resolved before we merge this in.

File: src/macad_gym/carla/multi_env.py
Comment (johnMinelli) [reply src/macad_gym/carla/multi_env.py:394]: To make it user configurable with abilty to connect from the extern? I personally launch the execution on a cluster and use the port mapped to the extern to connect. I f you don't like it, i've no problem in making a rollback.

Comment (praveen-palanisamy) [reply src/macad_gym/carla/multi_env.py:394]: I see. Remote connections to the CARLA server launched through MACAD-Gym sounds like an unintended use case but if that's useful, it's okay to keep your changes since the default argument value of `0` has the same behaviour as before. Please update the comment to make it clear that only `port=0` will request the sys to provide a free port dynamically. Not any other value. 
Keep in mind that if the `port` argument value supplied to this method is already in use by some other process, this process will crash because that exception is not handled (using a try except statement).

Comment (praveen-palanisamy) [reply src/macad_gym/carla/multi_env.py:394]: The `port` is not currently exposed to the user for configuration. Even if you add this argument, how will a user change it?

Comment (johnMinelli) [reply src/macad_gym/carla/multi_env.py]: i extendend the class that again, i think is not the intended use you wanted for the library 😅. I added the docstring for the method in the last commit

**Comment by praveen-palanisamy** [comment]
@johnMinelli : Could you please resolve the comment above on the `ray` camera type?
If you add an example or at least a test case for that, it will be good for reference. 
We can merge this PR once the review comments are resolved.
Thank you for the contributions!

**Comment by johnMinelli** [comment]
oh sorry, I totally missed your reply

**Review by praveen-palanisamy — APPROVED** [review]

**Pull request merged**
